pkg/apiserver/authn.go

```diff
@@ -19,18 +19,69 @@ package apiserver
 import (
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/authenticator"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/authenticator/bearertoken"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/plugin/pkg/auth/authenticator/request/union"
+	"github.com/GoogleCloudPlatform/kubernetes/plugin/pkg/auth/authenticator/request/x509"
 	"github.com/GoogleCloudPlatform/kubernetes/plugin/pkg/auth/authenticator/token/tokenfile"
 )
 
 // NewAuthenticatorFromTokenFile returns an authenticator.Request or an error
 func NewAuthenticatorFromTokenFile(tokenAuthFile string) (authenticator.Request, error) {
-	var authenticator authenticator.Request
-	if len(tokenAuthFile) != 0 {
-		tokenAuthenticator, err := tokenfile.NewCSV(tokenAuthFile)
+	if len(tokenAuthFile) == 0 {
+		return nil, nil
+	}
+
+	tokenAuthenticator, err := tokenfile.NewCSV(tokenAuthFile)
+	if err != nil {
+		return nil, err
+	}
+
+	return bearertoken.New(tokenAuthenticator), nil
+}
+
+// NewAuthenticatorFromClientCAFile returns an authenticator.Request or an error
+func NewAuthenticatorFromClientCAFile(clientCAFile string) (authenticator.Request, error) {
+	if len(clientCAFile) == 0 {
+		return nil, nil
+	}
+
+	roots, err := util.CertPoolFromFile(clientCAFile)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := x509.DefaultVerifyOptions()
+	opts.Roots = roots
+
+	return x509.New(opts, x509.CommonNameUserConversion), nil
+}
+
+// NewAuthenticator returns an authenticator.Request or an error
+func NewAuthenticator(clientCAFile string, tokenFile string) (authenticator.Request, error) {
```

Comment:
There are no remaining uses (outside of this file) of NewAuthenticatorFromTokenFile so I'd suggest making it private. That would also allow  you to remove some of the (then redundant) checks on the input variables being non-empty.